replica/transition_test.go

```diff
@@ -1 +1,298 @@
+			default:
+				Expect("PreCommitted type").To(Equal(""),
+					"Type is: %T", tranType)
+			}
+		})
+		It("Third should have nothing", func() {
+			tran, ok := tb.Dequeue(0)
+			Expect(ok).To(Equal(false), "dequeued type %t", tran)
+		})
+	})
+	Context("When given Proposed -> PreCommitted -> PreCommitted", func() {
+		tb := NewTransitionBuffer()
+
+		tb.Enqueue(Proposed{Block: block.Genesis()})
+		precom := PreCommitted{}
+		precom.Polka.Height = 0
+		tb.Enqueue(precom)
+		tb.Enqueue(precom)
+		It("First and Second should be a PreCommitted", func() {
+			tran, ok := tb.Dequeue(0)
+			Expect(ok).To(Equal(true))
+			switch tranType := tran.(type) {
+			case PreCommitted:
+			default:
+				Expect("PreCommitted type").To(Equal(""),
+					"Type is: %T", tranType)
+			}
+			tran, ok = tb.Dequeue(0)
+			Expect(ok).To(Equal(true))
+			switch tranType := tran.(type) {
+			case PreCommitted:
+			default:
+				Expect("PreCommitted type").To(Equal(""),
+					"Type is: %T", tranType)
+			}
+		})
+		It("Fourth should have nothing", func() {
+			tran, ok := tb.Dequeue(0)
+			Expect(ok).To(Equal(false), "dequeued type %t", tran)
+		})
+	})
+	Context("When given Transitions in incorrect order, panics", func() {
+		It("Should panic when PreCommitted -> Proposed", func() {
+			tb := NewTransitionBuffer()
+
+			precom := PreCommitted{}
+			precom.Polka.Height = 0
+			tb.Enqueue(precom)
+			Expect(func() {
+				tb.Enqueue(Proposed{Block: block.Genesis()})
+			}).Should(Panic())
+			_, ok := tb.Dequeue(0)
+			Expect(ok).To(Equal(true))
+			tran, ok := tb.Dequeue(0)
+			Expect(ok).To(Equal(false), "dequeued type %T", tran)
+		})
+		It("Should panic when PreVoted -> Proposed", func() {
+			tb := NewTransitionBuffer()
+
+			prevote := PreVoted{}
+			prevote.Height = 0
+			tb.Enqueue(prevote)
+			Expect(func() {
+				tb.Enqueue(Proposed{Block: block.Genesis()})
+			}).Should(Panic())
+			_, ok := tb.Dequeue(0)
+			Expect(ok).To(Equal(true))
+			tran, ok := tb.Dequeue(0)
+			Expect(ok).To(Equal(false), "dequeued type %T", tran)
+		})
+		It("Should panic when PreCommitted -> PreVoted", func() {
+			tb := NewTransitionBuffer()
+
+			precom := PreCommitted{}
+			precom.Polka.Height = 0
+			tb.Enqueue(precom)
+			Expect(func() {
+				prevote := PreVoted{}
+				prevote.Height = 0
+				tb.Enqueue(prevote)
+			}).Should(Panic())
+			_, ok := tb.Dequeue(0)
+			Expect(ok).To(Equal(true))
+			tran, ok := tb.Dequeue(0)
+			Expect(ok).To(Equal(false), "dequeued type %T", tran)
+		})
+	})
+	Context("When Drop is called", func() {
+		It("Should remove everything below the given height", func() {
+			tb := NewTransitionBuffer()
+
+			precom := PreCommitted{}
+			precom.Polka.Height = 0
+			tb.Enqueue(precom)
+			tb.Enqueue(precom)
+			precom.Polka.Height = 1
+			tb.Enqueue(precom)
+			tb.Drop(1)
+			tran, ok := tb.Dequeue(0)
+			Expect(ok).To(Equal(false), "dequeued type %T", tran)
+			_, ok = tb.Dequeue(1)
+			Expect(ok).To(Equal(true))
+		})
+	})
+})
+
+// type mockInput struct {
```

Comment:
If you don't need it, delete it.

Reply:
I was thinking of using it to provide fuzzy inputs in other testing modules, but considering the fact that it was designed with sequential Transitions (Propose -> PreVote -> etc..), I need to re-write it anyways.